virtual_rainforest/core/config.py

```diff
@@ -0,0 +1,362 @@
+                log_and_raise(
+                    f"Configuration file {file} is incorrectly formatted. Failed with "
+                    f"the following message:\n{err}",
+                    RuntimeError,
+                )
+
+    # Check if any tags are repeated across files
+    if len(conflicts) != 0:
+        # If so generate full list of errors
+        msg = "The following tags are defined in multiple config files:\n"
+        for conf in conflicts:
+            msg += f"{conf[0]} defined in both {conf[1]} and {conf[2]}\n"
+        msg = msg[:-1]
+        log_and_raise(msg, RuntimeError)
+
+    # Merge everything into a single dictionary
+    config_dict: dict = {}
+    for item in file_data:
+        dpath.util.merge(config_dict, item[1])
+
+    return config_dict
+
+
+def find_schema(config_dict: dict) -> list[str]:
+    """Find which schema the configuration requires to be loaded.
+
+    Args:
+        config_dict: The complete configuration settings for the particular model
+            instance
+    """
+
+    # Find which other schema should be searched for
+    try:
+        modules = config_dict["core"]["modules"]
+    except KeyError:
+        log_and_raise(
+            "Core configuration does not specify which other modules should be "
+            "configured!",
+            KeyError,
+        )
+
+    # Add core to list of modules if its not already included
+    if "core" not in modules:
+        modules.append("core")
+
+    if len(modules) != len(set(modules)):
+        log_and_raise(
+            f"The list of modules to configure given in the core configuration file "
+            f"repeats {len(modules) - len(set(modules))} names!",
+            RuntimeError,
+        )
+
+    return modules
+
+
+def construct_combined_schema(modules: list[str]) -> dict:
+    """Load validation schema for desired modules, and combine into a single schema.
+
+    Args:
+        modules: List of modules to load schema for
+    """
+
+    # Construct combined schema for all relevant modules
+    comb_schema: dict = {}
+
+    # Loop over expected modules and add them to the registry
+    for module in modules:
+        if module in SCHEMA_REGISTRY:
+            try:
+                m_schema = SCHEMA_REGISTRY[module]["properties"][module]
+            except KeyError as err:
+                log_and_raise(
+                    f"Schema for {module} module incorrectly structured, {err} key "
+                    f"missing!",
+                    KeyError,
+                )
+            # Store complete schema if no previous schema has been added
+            if comb_schema == {}:
+                comb_schema = SCHEMA_REGISTRY[module]
+            # Otherwise only save truncated part of the schema
+            else:
+                comb_schema["properties"][module] = m_schema
+                # Add module name to list of required modules
+                try:
+                    comb_schema["required"].append(module)
```

Comment:
If `comb_schema["required"]` is a list, I don't see how this might fail. I don't think a `try... except` block is needed.

Reply:
At the moment no check is performed that the "required" tag is actually included in a schema, so if the first schema is missing it then this step will fail. This might be better handled by a check that the "required" key is included in every schema actually, particularly as schema loading order changing could errors to suddenly appear with the current setup